util/fipstools/acvp/modulewrapper/modulewrapper.cc

```diff
@@ -3072,6 +3132,171 @@ static bool ED25519SigVer(const Span<const uint8_t> args[],
   return write_reply({Span<const uint8_t>(reply)});
 }
 
+template <int nid>
+static bool ML_DSA_KEYGEN(const Span<const uint8_t> args[],
+                          ReplyCallback write_reply) {
+  const Span<const uint8_t> seed = args[0];
+
+  //init params of the correct size based on provided nid
+  ml_dsa_params params;
+  if (nid == NID_MLDSA44) {
+    ml_dsa_44_params_init(&params);
+  }
+  else if (nid == NID_MLDSA65) {
+    ml_dsa_65_params_init(&params);
+  }
+  else if (nid == NID_MLDSA87) {
+    ml_dsa_87_params_init(&params);
+  }
+
+  // create public and private key buffers
+  std::vector<uint8_t> public_key(params.public_key_bytes);
+  std::vector<uint8_t> private_key(params.secret_key_bytes);
+
+  // generate the keys
+  if (nid == NID_MLDSA44) {
+    if (!ml_dsa_44_keypair_internal(public_key.data(), private_key.data(), seed.data())) {
+      return false;
+    }
+  }
+  else if (nid == NID_MLDSA65) {
+    if (!ml_dsa_65_keypair_internal(public_key.data(), private_key.data(), seed.data())) {
+      return false;
+    }
+  }
+  else if (nid == NID_MLDSA87) {
+    if (!ml_dsa_87_keypair_internal(public_key.data(), private_key.data(), seed.data())) {
+      return false;
+    }
+  }
+  return write_reply({Span<const uint8_t>(public_key.data(), public_key.size()),
+                      Span<const uint8_t>(private_key.data(), private_key.size())});
+}
+
+template <int nid>
+static bool ML_DSA_SIGGEN(const Span<const uint8_t> args[],
+                         ReplyCallback write_reply) {
+  const Span<const uint8_t> sk = args[0];
+  const Span<const uint8_t> msg = args[1];
+  const Span<const uint8_t> mu = args[2];
+  const Span<const uint8_t> rnd = args[3];
+  const Span<const uint8_t> extmu = args[4];
+
+  ml_dsa_params params;
+  if (nid == NID_MLDSA44) {
+    ml_dsa_44_params_init(&params);
+  }
+  else if (nid == NID_MLDSA65) {
+    ml_dsa_65_params_init(&params);
+  }
+  else if (nid == NID_MLDSA87) {
+    ml_dsa_87_params_init(&params);
+  }
+
+  size_t signature_len = params.bytes;
+  std::vector<uint8_t> signature(signature_len);
+
+  // generate the signatures raw sign mode
+  if (extmu.data()[0] == 0) {
+    if (nid == NID_MLDSA44) {
+      if (!ml_dsa_44_sign_internal(sk.data(), signature.data(), &signature_len,
+                                   msg.data(), msg.size(), nullptr, 0, rnd.data())) {
+        return false;
+      }
+    }
+    else if (nid == NID_MLDSA65) {
+      if (!ml_dsa_65_sign_internal(sk.data(), signature.data(), &signature_len,
+                                   msg.data(), msg.size(), nullptr, 0, rnd.data())) {
+        return false;
+      }
+    }
+    else if (nid == NID_MLDSA87) {
+      if (!ml_dsa_87_sign_internal(sk.data(), signature.data(), &signature_len,
+                                   msg.data(), msg.size(), nullptr, 0, rnd.data())) {
+        return false;
+      }
+    }
+  }
```

Comment:
This would all happen automatically if you used the nice EVP API you just added.

Reply:
See comment above

Reply:
These internal functions are designed for use in ACVP testing, and thus, we must use them.